File: app/src/main/java/com/huanchengfly/tieba/post/activities/UserActivity.java
package com.huanchengfly.tieba.post.activities;

import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.tabs.TabLayout;
import com.huanchengfly.tieba.post.ui.theme.utils.ThemeUtils;
import com.huanchengfly.tieba.post.api.TiebaApi;
import com.huanchengfly.tieba.post.api.models.CommonResponse;
import com.huanchengfly.tieba.post.api.models.ProfileBean;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.adapters.FragmentTabViewPagerAdapter;
import com.huanchengfly.tieba.post.fragments.UserLikeForumFragment;
import com.huanchengfly.tieba.post.fragments.UserPostFragment;
import com.huanchengfly.tieba.post.models.PhotoViewBean;
import com.huanchengfly.tieba.post.models.database.Account;
import com.huanchengfly.tieba.post.models.database.Block;
import com.huanchengfly.tieba.post.utils.AccountUtil;
import com.huanchengfly.tieba.post.utils.ImageUtil;
import com.huanchengfly.tieba.post.utils.ThemeUtil;
import com.huanchengfly.tieba.post.widgets.theme.TintMaterialButton;

import org.jetbrains.annotations.NotNull;

import butterknife.BindView;
import butterknife.OnClick;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.huanchengfly.tieba.post.utils.Util.changeAlpha;

public class UserActivity extends BaseActivity {
    public static final String TAG = "UserActivity";
    public static final String EXTRA_UID = "uid";
    public static final String EXTRA_TAB = "tab";
    public static final String EXTRA_AVATAR = "avatar";

    public static final int TAB_THREAD = 0;
    public static final int TAB_REPLY = 1;
    public static final int TAB_LIKE_FORUM = 2;

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.user_center_avatar)
    ImageView avatarView;
    @BindView(R.id.title_view)
    TextView titleView;
    @BindView(R.id.user_center_stat)
    TextView statView;
    @BindView(R.id.user_center_action_btn)
    TintMaterialButton actionBtn;
    @BindView(R.id.loading_view)
    View loadingView;

    private ProfileBean profileBean;

    private String uid;
    private int tab;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_user;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ThemeUtil.setTranslucentThemeBackground(findViewById(R.id.background));
        uid = getIntent().getStringExtra(EXTRA_UID);
        tab = getIntent().getIntExtra(EXTRA_TAB, TAB_THREAD);
        String avatar = getIntent().getStringExtra(EXTRA_AVATAR);
        if (uid == null) {
            finish();
            return;
        }
        FragmentTabViewPagerAdapter adapter = new FragmentTabViewPagerAdapter(getSupportFragmentManager());
        ViewPager viewPager = (ViewPager) findViewById(R.id.user_center_vp);
        TabLayout tabLayout = (TabLayout) findViewById(R.id.user_center_tab);
        AppBarLayout appBarLayout = (AppBarLayout) findViewById(R.id.appbar);
        actionBtn.setVisibility(View.GONE);
        if (!TextUtils.isEmpty(avatar)) {
            loadingView.setVisibility(View.GONE);
            ImageUtil.load(avatarView, ImageUtil.LOAD_TYPE_AVATAR, avatar);
            ImageUtil.initImageView(avatarView, new PhotoViewBean(avatar));
        }
        appBarLayout.addOnOffsetChangedListener((appBarLayout1, verticalOffset) -> {
            toolbar.setBackgroundColor(changeAlpha(ThemeUtils.getColorByAttr(this, R.attr.colorToolbar), Math.abs(verticalOffset * 1.0f) / appBarLayout1.getTotalScrollRange()));
            if (profileBean != null && profileBean.getUser() != null && Math.abs(verticalOffset) >= appBarLayout1.getTotalScrollRange()) {
                toolbar.setTitle(profileBean.getUser().getNameShow());
            } else {
                toolbar.setTitle(null);
            }
        });
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        viewPager.setAdapter(adapter);
        viewPager.setOffscreenPageLimit(3);
        tabLayout.setupWithViewPager(viewPager);
        TiebaApi.getInstance().profile(uid).enqueue(new Callback<ProfileBean>() {
            @Override
            public void onResponse(@NotNull Call<ProfileBean> call, @NotNull Response<ProfileBean> response) {
                ProfileBean data = response.body();
                actionBtn.setVisibility(View.VISIBLE);
                loadingView.setVisibility(View.GONE);
                View dividerView = findViewById(R.id.user_center_divider);
                if (ThemeUtils.getColorByAttr(UserActivity.this, R.attr.colorBg) == ThemeUtils.getColorByAttr(UserActivity.this, R.attr.colorToolbar)) {
                    dividerView.setVisibility(View.VISIBLE);
                }
                if (ThemeUtils.getColorByAttr(UserActivity.this, R.attr.colorToolbar) == ThemeUtils.getColorByAttr(UserActivity.this, R.attr.colorAccent)) {
                    actionBtn.setTextColor(ColorStateList.valueOf(Color.WHITE));
                    actionBtn.setStrokeColor(ColorStateList.valueOf(Color.WHITE));
                }
                profileBean = data;
                refreshHeader();
                adapter.clear();
                adapter.addFragment(UserPostFragment.newInstance(uid, true), "贴子 " + data.getUser().getThreadNum());
                adapter.addFragment(UserPostFragment.newInstance(uid, false), "回复 " + data.getUser().getRepostNum());
                adapter.addFragment(UserLikeForumFragment.newInstance(uid), "关注吧 " + data.getUser().getMyLikeNum());
                viewPager.setCurrentItem(tab, false);
            }

            @Override
            public void onFailure(@NotNull Call<ProfileBean> call, @NotNull Throwable t) {
            }
        });
    }

    public void refreshHeader() {
        titleView.setText(profileBean.getUser().getNameShow());
        statView.setText(getString(R.string.tip_stat, profileBean.getUser().getConcernNum(), profileBean.getUser().getFansNum()));
        if (avatarView.getTag() == null) {
            ImageUtil.load(avatarView, ImageUtil.LOAD_TYPE_AVATAR, "http://tb.himg.baidu.com/sys/portrait/item/" + profileBean.getUser().getPortrait());
            ImageUtil.initImageView(avatarView, new PhotoViewBean("http://tb.himg.baidu.com/sys/portrait/item/" + profileBean.getUser().getPortrait()));
        }
        if (TextUtils.equals(AccountUtil.getUid(this), profileBean.getUser().getId())) {
            actionBtn.setText(R.string.menu_edit_info);
        } else {
            if ("1".equals(profileBean.getUser().getHasConcerned())) {
                actionBtn.setText(R.string.button_unfollow);
            } else {
                actionBtn.setText(R.string.button_follow);
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_user_space, menu);
        Account account = AccountUtil.getLoginInfo(this);
        if (account != null && TextUtils.equals(account.getUid(), uid)) {
            menu.findItem(R.id.menu_block).setVisible(false);
            menu.findItem(R.id.menu_edit_info).setVisible(true);
        } else {
            menu.findItem(R.id.menu_block).setVisible(true);
            menu.findItem(R.id.menu_edit_info).setVisible(false);
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_block_black:
            case R.id.menu_block_white:
                int category = item.getItemId() == R.id.menu_block_black ? Block.CATEGORY_BLACK_LIST : Block.CATEGORY_WHITE_LIST;
                new Block()
                        .setUid(profileBean.getUser().getId())
                        .setUsername(profileBean.getUser().getName())
                        .setType(Block.TYPE_USER)
                        .setCategory(category)
                        .saveAsync()
                        .listen(success -> {
                            if (success) {
                                Toast.makeText(this, R.string.toast_add_success, Toast.LENGTH_SHORT).show();
                            }
                        });
                return true;
            case R.id.menu_edit_info:
                startActivity(WebViewActivity.newIntent(this, getString(R.string.url_edit_info)));
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.user_center_action_btn)
    public void onActionBtnClick(View view) {
        if (TextUtils.equals(profileBean.getUser().getId(), AccountUtil.getUid(this))) {
            startActivity(WebViewActivity.newIntent(this, getString(R.string.url_edit_info)));
            return;
        }
        if ("1".equals(profileBean.getUser().getHasConcerned())) {
            TiebaApi.getInstance().unfollow(profileBean.getUser().getPortrait(), AccountUtil.getLoginInfo(this).getTbs()).enqueue(new Callback<CommonResponse>() {
                @Override
                public void onResponse(@NotNull Call<CommonResponse> call, @NotNull Response<CommonResponse> response) {
                    CommonResponse data = response.body();
                    Toast.makeText(UserActivity.this, data.getErrorMsg(), Toast.LENGTH_SHORT).show();
                    profileBean.getUser().setHasConcerned("0");
                    refreshHeader();
                }

                @Override
                public void onFailure(@NotNull Call<CommonResponse> call, @NotNull Throwable t) {
                    Toast.makeText(UserActivity.this, t.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
        } else {
            TiebaApi.getInstance().follow(profileBean.getUser().getPortrait(), AccountUtil.getLoginInfo(this).getTbs()).enqueue(new Callback<CommonResponse>() {
                @Override
                public void onResponse(@NotNull Call<CommonResponse> call, @NotNull Response<CommonResponse> response) {
                    CommonResponse data = response.body();
                    Toast.makeText(UserActivity.this, data.getErrorMsg(), Toast.LENGTH_SHORT).show();
                    profileBean.getUser().setHasConcerned("1");
                    refreshHeader();
                }

                @Override
                public void onFailure(@NotNull Call<CommonResponse> call, @NotNull Throwable t) {
                    Toast.makeText(UserActivity.this, t.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/activities/UserActivity.kt
@file:SuppressLint("NonConstantResourceId")

package com.huanchengfly.tieba.post.activities

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.os.Bundle
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import butterknife.OnClick
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.google.android.material.tabs.TabLayout
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.adapters.FragmentTabViewPagerAdapter
import com.huanchengfly.tieba.post.api.TiebaApi
import com.huanchengfly.tieba.post.api.models.protos.profile.ProfileResponseData
import com.huanchengfly.tieba.post.api.retrofit.exception.getErrorMessage
import com.huanchengfly.tieba.post.fragments.UserLikeForumFragment
import com.huanchengfly.tieba.post.fragments.UserPostFragment
import com.huanchengfly.tieba.post.goToActivity
import com.huanchengfly.tieba.post.models.PhotoViewBean
import com.huanchengfly.tieba.post.models.database.Block
import com.huanchengfly.tieba.post.plugins.PluginManager
import com.huanchengfly.tieba.post.toastShort
import com.huanchengfly.tieba.post.ui.page.editprofile.view.EditProfileActivity
import com.huanchengfly.tieba.post.ui.widgets.theme.TintMaterialButton
import com.huanchengfly.tieba.post.ui.widgets.theme.TintToolbar
import com.huanchengfly.tieba.post.utils.*
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.launch
import kotlin.math.abs


class UserActivity : BaseActivity() {
    @BindView(R.id.toolbar)
    lateinit var toolbar: TintToolbar

    @BindView(R.id.appbar)
    lateinit var appbar: AppBarLayout

    @BindView(R.id.user_center_avatar)
    lateinit var avatarView: ImageView

    @BindView(R.id.title_view)
    lateinit var titleView: TextView

    @BindView(R.id.user_center_slogan)
    lateinit var sloganView: TextView

    @BindView(R.id.user_center_stat_follow)
    lateinit var followStatTv: TextView

    @BindView(R.id.user_center_stat_fans)
    lateinit var fansStatTv: TextView

    @BindView(R.id.user_sex)
    lateinit var sexTv: TextView

    @BindView(R.id.user_tb_age)
    lateinit var tbAgeTv: TextView

    @BindView(R.id.user_center_action_btn)
    lateinit var actionBtn: TintMaterialButton

    @BindView(R.id.loading_view)
    lateinit var loadingView: View

    @BindView(R.id.fake_status_bar)
    lateinit var fakeStatusBarView: View

    @BindView(R.id.user_center_header)
    lateinit var headerView: View

    @BindView(R.id.user_center_header_mask)
    lateinit var headerMaskView: View

    @BindView(R.id.user_info_chips)
    lateinit var infoChips: View

    private var profileBean: ProfileResponseData? = null
    private var uid: String? = null
    private var tab = 0

    override fun getLayoutId(): Int {
        return R.layout.activity_user
    }

    override val isNeedImmersionBar: Boolean
        get() = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        StatusBarUtil.transparentStatusBar(this)
        (toolbar.layoutParams as CollapsingToolbarLayout.LayoutParams).topMargin =
            StatusBarUtil.getStatusBarHeight(this)
        fakeStatusBarView.minimumHeight = StatusBarUtil.getStatusBarHeight(this)
        ThemeUtil.setTranslucentThemeBackground(findViewById(R.id.background))
        uid = intent.getStringExtra(EXTRA_UID)
        tab = intent.getIntExtra(EXTRA_TAB, TAB_THREAD)
        val avatar = intent.getStringExtra(EXTRA_AVATAR)
        val uidLong = uid?.toLongOrNull()
        if (uidLong == null) {
            finish()
            return
        }
        val adapter = FragmentTabViewPagerAdapter(supportFragmentManager)
        val viewPager = findViewById<ViewPager>(R.id.user_center_vp)
        val tabLayout = findViewById<TabLayout>(R.id.user_center_tab)
        actionBtn.visibility = View.GONE
        if (!TextUtils.isEmpty(avatar)) {
            loadingView.visibility = View.GONE
            ImageUtil.load(avatarView, ImageUtil.LOAD_TYPE_AVATAR, StringUtil.getAvatarUrl(avatar))
            ImageUtil.initImageView(
                avatarView,
                PhotoViewBean(StringUtil.getAvatarUrl(avatar), StringUtil.getBigAvatarUrl(avatar))
            )
        }
        appbar.addOnOffsetChangedListener { appBarLayout: AppBarLayout, verticalOffset: Int ->
            val percent = abs(verticalOffset * 1.0f) / appBarLayout.totalScrollRange
            headerView.alpha = 1f - percent
            headerMaskView.alpha = percent
            if (profileBean != null && abs(verticalOffset) >= appBarLayout.totalScrollRange) {
                toolbar.title = profileBean?.user?.nameShow
            } else {
                toolbar.title = null
            }
        }
        viewPager.adapter = adapter
        viewPager.offscreenPageLimit = 3
        tabLayout.setupWithViewPager(viewPager)
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        launch {
            TiebaApi.getInstance()
                .userProfileFlow(uidLong)
                .catch { toastShort(it.getErrorMessage()) }
                .collect {
                    actionBtn.visibility = View.VISIBLE
                    loadingView.visibility = View.GONE
                    val data = it.data_
                    if (data?.user == null) {
                        return@collect
                    }
                    profileBean = data
                    refreshHeader()
                    adapter.clear()
                    adapter.addFragment(
                        UserPostFragment.newInstance(uid, true),
                        "贴子 " + data.user.thread_num
                    )
                    adapter.addFragment(
                        UserPostFragment.newInstance(uid, false),
                        "回复 " + data.user.post_num
                    )
                    adapter.addFragment(
                        UserLikeForumFragment.newInstance(uid),
                        "关注吧 " + data.user.my_like_num
                    )
                    viewPager.setCurrentItem(tab, false)
                }
        }
        listOf(
            followStatTv,
            fansStatTv
        ).forEach {
            it.typeface = Typeface.createFromAsset(assets, "bebas.ttf")
        }
    }

    private fun refreshHeader() {
        profileBean?.user?.let {
            titleView.text = it.nameShow
            sloganView.text = it.intro
            followStatTv.text = "${it.concern_num}"
            fansStatTv.text = "${it.fans_num}"
            //getString(R.string.tip_stat, profileBean!!.user!!.concernNum, profileBean!!.user!!.fansNum)
            if (avatarView.tag == null) {
                ImageUtil.load(
                    avatarView,
                    ImageUtil.LOAD_TYPE_AVATAR,
                    StringUtil.getAvatarUrl(it.portrait)
                )
                ImageUtil.initImageView(
                    avatarView,
                    PhotoViewBean(
                        StringUtil.getBigAvatarUrl(it.portraith),
                        StringUtil.getBigAvatarUrl(it.portraith)
                    )
                )
            }
            if (TextUtils.equals(AccountUtil.getUid(), "${it.id}")) {
                actionBtn.setText(R.string.menu_edit_info)
            } else {
                if (1 == it.has_concerned) {
                    actionBtn.setText(R.string.button_unfollow)
                } else {
                    actionBtn.setText(R.string.button_follow)
                }
            }
            sexTv.text =
                if (it.sex == 1) "♂" else if (it.sex == 2) "♀" else "?"
            tbAgeTv.text = getString(R.string.tb_age, it.tb_age)
            infoChips.visibility = View.VISIBLE
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_user_space, menu)
        val account = AccountUtil.getLoginInfo()
        if (account != null && TextUtils.equals(account.uid, uid)) {
            menu.findItem(R.id.menu_block).isVisible = false
            menu.findItem(R.id.menu_edit_info).isVisible = true
        } else {
            menu.findItem(R.id.menu_block).isVisible = true
            menu.findItem(R.id.menu_edit_info).isVisible = false
        }
        PluginManager.initPluginMenu(menu, PluginManager.MENU_USER_ACTIVITY)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_block_black, R.id.menu_block_white -> {
                profileBean?.user?.let { user ->
                    val category =
                        if (item.itemId == R.id.menu_block_black) Block.CATEGORY_BLACK_LIST else Block.CATEGORY_WHITE_LIST
                    BlockManager.addBlockAsync(
                        Block(
                            category = category,
                            type = Block.TYPE_USER,
                            username = user.name,
                            uid = user.id.toString()
                        )
                    ) {
                        if (it) {
                            toastShort(R.string.toast_add_success)
                        }
                    }
                }
                return true
            }

            R.id.menu_edit_info -> {
                startActivity(WebViewActivity.newIntent(this, getString(R.string.url_edit_info)))
                return true
            }
        }
        return if (PluginManager.performPluginMenuClick(
                PluginManager.MENU_USER_ACTIVITY,
                item.itemId,
                profileBean
            )
        ) {
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    @OnClick(R.id.user_center_action_btn)
    fun onActionBtnClick(view: View?) {
        profileBean?.user?.let { user ->
            if (TextUtils.equals(user.id.toString(), AccountUtil.getUid())) {
                goToActivity<EditProfileActivity>()
                return
            }
            if (1 == user.has_concerned) {
                launch {
                    TiebaApi.getInstance()
                        .unfollowFlow(
                            user.portrait,
                            AccountUtil.getLoginInfo()!!.tbs
                        )
                        .catch { e ->
                            toastShort(e.getErrorMessage())
                        }
                        .collect {
                            profileBean = profileBean!!.copy(
                                user = profileBean?.user!!.copy(
                                    has_concerned = 0
                                )
                            )
                            refreshHeader()
                        }
                }
            } else {
                launch {
                    TiebaApi.getInstance()
                        .followFlow(
                            user.portrait,
                            AccountUtil.getLoginInfo()!!.tbs
                        )
                        .catch { e ->
                            toastShort(e.getErrorMessage())
                        }
                        .collect {
                            if ("1" == it.info.isToast) {
                                toastShort(it.info.toastText)
                            }
                            profileBean = profileBean!!.copy(
                                user = profileBean?.user!!.copy(
                                    has_concerned = 1
                                )
                            )
                            refreshHeader()
                        }
                }
            }
        }
    }

    companion object {
        const val TAG = "UserActivity"
        const val EXTRA_UID = "uid"
        const val EXTRA_TAB = "tab"
        const val EXTRA_AVATAR = "avatar"
        const val TAB_THREAD = 0
        const val TAB_REPLY = 1
        const val TAB_LIKE_FORUM = 2

        @JvmOverloads
        fun launch(
            context: Context,
            userId: String,
            avatarUrl: String? = null,
        ) {
            context.goToActivity<UserActivity> {
                putExtra(EXTRA_UID, userId)
                putExtra(EXTRA_AVATAR, avatarUrl)
            }
        }
    }
}